File: python/Q4_median_of_two_sorted_arrays.py
#
# @lc app=leetcode id=4 lang=python3
#
# [4] Median of Two Sorted Arrays
#


# @lc code=start
class Solution:
    def findMedianSortedArrays(self, nums1: list[int], nums2: list[int]) -> float:
        short: list[int] = nums1 if len(nums1) <= len(nums2) else nums2
        long: list[int] = nums2 if len(nums1) <= len(nums2) else nums1

        total: int = len(short) + len(long)
        half: int = total // 2

        left: int = 0
        right: int = len(short)
        result: float = 0.0

        while left <= right:
            mid: int = left + ((right - left) // 2)
            other_mid: int = half - mid

            short_left: int = short[mid - 1] if mid > 0 else -(10**6) - 1
            short_right: int = short[mid] if mid < len(short) else (10**6) + 1
            long_left: int = long[other_mid - 1] if other_mid > 0 else -(10**6) - 1
            long_right: int = long[other_mid] if other_mid < len(long) else (10**6) + 1

            if short_left <= long_right and long_left <= short_right:
                max_left: int = max(short_left, long_left)
                min_right: int = min(short_right, long_right)

                if total % 2 != 0:
                    result = min_right
                else:
                    result = (max_left + min_right) / 2.0

                break

            if short_left > long_right:
                right = mid - 1
            else:
                left = mid + 1

        return result


# @lc code=end

File: rust/src/leetcode/Q4_median_of_two_sorted_arrays.rs
/*
 * @lc app=leetcode id=4 lang=rust
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start
use std::cmp::{max, min};

impl Solution {
    fn order_list_by_length(nums1: Vec<i32>, nums2: Vec<i32>) -> (Vec<i32>, Vec<i32>) {
        if nums1.len() <= nums2.len() {
            (nums1, nums2)
        } else {
            (nums2, nums1)
        }
    }

    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let (shorter, longer) = Self::order_list_by_length(nums1, nums2);

        let half_length: i32 = (shorter.len() + longer.len()) as i32 / 2;

        let mut left: i32 = 0;
        let mut right: i32 = shorter.len() as i32;

        while left <= right {
            let shorter_mid: i32 = left + (right - left) / 2;
            let longer_mid: i32 = half_length - shorter_mid;

            let shorter_left: i32 = if shorter_mid != 0 {
                shorter[shorter_mid as usize - 1]
            } else {
                i32::MIN
            };
            let shorter_right: i32 = if shorter_mid != shorter.len() as i32 {
                shorter[shorter_mid as usize]
            } else {
                i32::MAX
            };
            let longer_left: i32 = if longer_mid != 0 {
                longer[longer_mid as usize - 1]
            } else {
                i32::MIN
            };
            let longer_right: i32 = if longer_mid != longer.len() as i32 {
                longer[longer_mid as usize]
            } else {
                i32::MAX
            };

            if shorter_left <= longer_right && longer_left <= shorter_right {
                if (shorter.len() + longer.len()) % 2 == 1 {
                    return min(shorter_right, longer_right) as f64;
                } else {
                    let median_left: i32 = max(shorter_left, longer_left);
                    let median_right: i32 = min(shorter_right, longer_right);

                    return (median_left + median_right) as f64 / 2.0;
                }
            }

            if shorter_left > longer_right {
                right = shorter_mid - 1;
            } else {
                left = shorter_mid + 1;
            }
        }

        0.0
    }
}
// @lc code=end
